### ISSUE TITLE
How to connect CloudyKit/jet to web framework gin?

### ISSUE THREAD

**Issue by Gipi66** [CloudyKit/jet#51]
I work whith gin and want connect in  project templates engine CloudyKit/jet.
When you try to connect i have error becouse jet use dont standart type for templates (
if I am not mistaken *jet.Templates)

How to connect CloudyKit/jet to web framework gin?

**Comment by annismckenzie** [comment]
I'll write something up for you in an hour – we use that combination in production. :)

**Comment by annismckenzie** [comment]
Take a look at [Rendering templates](https://github.com/CloudyKit/jet/wiki/Rendering-templates) in the wiki. The only thing that's missing for you to make it work is to pass Gin's `Writer` on its context to the templates `Execute` function – here's a short example using one of their examples:
```go
func main() {
    router := gin.Default()

    // This handler will match /user/john but will not match neither /user/ or /user
    router.GET("/user/:name", func(c *gin.Context) {
        name := c.Param("name")
        vars := make(jet.VarMap)
        vars.Set("name", name)
        var View = jet.NewHTMLSet("./views"))
        t, err := View.GetTemplate("your_template.jet")
        if err != nil {
            // template could not be loaded, log the error?
        }
        c.Writer.WriteHeader(200)
        if err = t.Execute(c.Writer, vars, nil); err != nil {
            // log the error
	}
    })

    router.Run(":8080")
}
```

I imagine that you'd put the setup of the `jet.Set` to a more central place within your app, I put it in there for the sake of completeness. As with all other web frameworks you still need to handle 404s, 500s or template rendering errors in some other way – we're currently doing this in a middleware where we switch out the `c.Writer` before passing it down the middleware chain and on to the controller and decide after the controller action ran whether to return it as the response or do something different (setting the HTTP status code to 500 etc.).

I hope that you can glean the necessary information from this example. You need to understand that you can't use Gin's other response mechanisms like `c.String`, `c.JSON`, `c.HTML` or `c.YAML` because these don't render templates but render directly into the response writer (the same as jet does).

**Comment by jerichohuang** [comment]
Hi @annismckenzie ,

I met the same problem of binding gin and jet in my project. I use the same way as you showed in your example above to bind them.

You also mentioned that you created a middleware to handler the `c.Writer` in your project and I think that is a good way to do it. But I am still not very clear about how you implemented your middleware. 

Could you please kindly give more information or example about that part?

**Comment by annismckenzie** [comment]
Sorry for being unresponsive @jerichohuang.

That middleware I was talking about is more of a yak shave – I'm not sure you'd like to use it. It's part of our session handling so it's also there to make changing the session possible throughout the continued middleware chain and the controller action, all without one change having to be persisted to the session cookie by user code. With that said, you also need a special response writer and I'll paste that as well: here's the [gist](https://gist.github.com/annismckenzie/8d1586e31cfcd4edc5e00ad5b6397519).

To sum up: it's far from efficient when it comes to buffers and copying data around. For us it's not a problem because in the grand scheme of handling a request the runtime in microseconds, allocs and bytes pale in comparison to the rest of the code so keep that in mind.

**Comment by annismckenzie** [comment]
Closing because there hasn't been a response from the OP (here and on Gitter).
